torch/distributed/rpc/api.py

```diff
@@ -113,17 +173,22 @@ def shutdown(graceful=True):
         >>> rpc.shutdown()
     """
     global _agent
-    if _agent:
-        if graceful:
-            _wait_all_workers()
-        _destroy_rref_context(_ignore_rref_leak)
-        _agent.shutdown()
-        # clean up python rpc handler in shutdown(), see comments in
-        # PythonRpcHandler::cleanup(), call it in python API because the
-        # cleanup() function has python dependency, it assumes python
-        # interpreter exists
-        _cleanup_python_rpc_handler()
-        _agent = None
+
+    if _agent is None:
+        return
+
+    if graceful:
+        _wait_all_workers()
+        _agent.join()
```

Comment:
Do we need to do this? This is not implemented for all agents?

Reply:
There is a RRef leak issue we need to fix before we can remove this line.

Two steps,

- Make all nexted RPC calls wait on futures they created so that they are chained.
- Add a API to RRefContext for proactively cleaning up local forks. (I know how to do this, I can help here. Will coordinate with @mrshenli )

Reply:
Could you elaborate what the RRef leak issue is and why we don't see it in other agents where the join() is pretty much a no-op.

Reply:
@pritamdamania 
The problem is now we depends on Python GC whose timing we don't have control on to send out delete fork messages.
It looks good now. 
Because in other agent, we sleep for 2 seconds, which is sufficient for 1) Python interpret to trigger GC and 2) the delete fork messages are flushed.

Reply:
Hmm, can we create a github issue describing the problem and the plan to fix it? We shouldn't have this 2 second sleep long term.

Reply:
@pritamdamania87 

Created.
https://github.com/pytorch/pytorch/issues/31325